compiler/src/iree/compiler/Preprocessing/Common/ConvertConvToChannelsLast.cpp

```diff
@@ -0,0 +1,886 @@
+                     AffineMap inputMap, AffineMap filterMap,
+                     AffineMap outputMap, SmallVector<unsigned> newDimOrder,
+                     SmallVector<utils::IteratorType> newIteratorTypes) {
+  AffineMap newInputMap = inputMap;
+  AffineMap newFilterMap = filterMap;
+  AffineMap newOutputMap = outputMap;
+  if (!newDimOrder.empty()) {
+    DenseMap<AffineExpr, AffineExpr> dimMap;
+    for (auto [newDim, oldDim] : llvm::enumerate(newDimOrder)) {
+      dimMap[b.getAffineDimExpr(oldDim)] = b.getAffineDimExpr(newDim);
+    }
+    newInputMap = inputMap.replace(dimMap,
+                                   /*numResultDims=*/newDimOrder.size(),
+                                   /*numResultSymbols=*/0);
+    newFilterMap = filterMap.replace(dimMap,
+                                     /*numResultDims=*/newDimOrder.size(),
+                                     /*numResultSymbols=*/0);
+    newOutputMap = outputMap.replace(dimMap,
+                                     /*numResultDims=*/newDimOrder.size(),
+                                     /*numResultSymbols=*/0);
+  }
+  SmallVector<utils::IteratorType> iterators = srcConv.getIteratorTypesArray();
+  iterators.append(newIteratorTypes);
+  auto genericConv = b.create<linalg::GenericOp>(
+      loc, output.getType(), ValueRange{input, filter}, output,
+      ArrayRef<AffineMap>{newInputMap, newFilterMap, newOutputMap}, iterators);
+  IRMapping mapper;
+  srcConv->getRegion(0).cloneInto(&genericConv.getRegion(), mapper);
+  return genericConv.getResult(0);
+}
+
+template <typename sourceNamedConvTy, typename targetNamedConvTy>
+static Value
+namedConvBuilderFn(OpBuilder &b, Location loc, linalg::LinalgOp srcConv,
+                   Value input, Value filter, Value output, AffineMap inputMap,
+                   AffineMap filterMap, AffineMap outputMap,
+                   SmallVector<unsigned> newDimOrder,
+                   SmallVector<utils::IteratorType> newIteratorTypes) {
+  sourceNamedConvTy namedConv = cast<sourceNamedConvTy>(srcConv);
+  return b
+      .create<targetNamedConvTy>(
+          loc, output.getType(), ValueRange{input, filter}, output,
+          namedConv.getStrides(), namedConv.getDilations())
+      .getResult(0);
+}
+
+// Normalizes the given permutation vector. Expects the input permutation is
+// shifted by a constant, for example, [4, 6, 5] -> [0, 2, 1].
+static TransposeIndices getNormalizedIndices(TransposeIndices targetIndices) {
+  int startDim = *std::min_element(targetIndices.begin(), targetIndices.end());
+  TransposeIndices normalized(targetIndices.size());
+  for (auto i : llvm::enumerate(targetIndices)) {
+    normalized[i.index()] = i.value() - startDim;
+  }
+  return normalized;
+}
+
+// Inverts the given shifted permutation vector. For example,
+// [2, 0, 1] + 4 -> [1, 2, 0] + 4
+// [6, 4, 5] -> [5, 6, 4]
+static TransposeIndices invertIndices(TransposeIndices targetIndices) {
```

Comment:
I think these methods are there upstream as well.

Reply:
These methods only exist for `[2, 0, 1] -> [1, 2, 0]` type of permutations (those that start at 0). The more accurate name for `TransposeIndices` is `InnerDimsPos` (literally directly what we use for the pack).